pkgs/test/security/check-quickfix.nix

```diff
@@ -0,0 +1,201 @@
+let
+  testDrv = stdenv: name: rec {
+      inherit name;
+      buildInputs = [ ];
+      buildCommand = ''
+        mkdir -p $out
+        touch $out/installed
+        echo ${name} >> $out/installed
+        touch $out/dependency
+        echo $out >> $out/dependency
+      '';
+
+      meta = with stdenv.lib; {
+        homepage = https://nixos.org/;
+        description = "Test case";
+        longDescription = "Test case";
+        license = licenses.mit;
+        maintainers = [ maintainers.pierron ];
+        platforms = platforms.all;
+      };
+    };
+
+  testDrvWithDep = stdenv: name: dep:
+    let drv = testDrv stdenv name; in drv // {
+      buildInputs = [ dep ];
+      buildCommand = drv.buildCommand + ''
+        cat ${dep}/installed >> $out/installed
+        cat ${dep}/dependency >> $out/dependency
+      '';
+    };
+
+  testPkg = stdenv: name:
+    stdenv.mkDerivation (testDrv stdenv name);
+
+  testPkgWithDep = stdenv: name: dep:
+    stdenv.mkDerivation (testDrvWithDep stdenv name dep);
+
+  pkgTestDead = { stdenv, name ? "dead-1.0.0" }:
+    testPkg stdenv name;
+  pkgTestBeef = { stdenv, name ? "beef-1.0.0", test-dead }:
+    testPkgWithDep stdenv name test-dead;
+  pkgTestAte = { stdenv, name ? "ate-1.0.0", test-beef }:
+    testPkgWithDep stdenv name test-beef;
+  pkgTestBad = { stdenv, name ? "bad-1.0.0", test-ate }:
+    testPkgWithDep stdenv name test-ate;
+  pkgTestFood = { stdenv, name ? "food-1.0.0", test-bad }:
+    testPkgWithDep stdenv name test-bad;
+
+  originalPackages = {
+    adapters = import ../../../pkgs/stdenv/adapters.nix;
+    builders = import ../../../pkgs/build-support/trivial-builders.nix;
+    stdenv = import ../../../pkgs/top-level/stdenv.nix;
+    all = import ../../../pkgs/top-level/all-packages.nix;
+    aliases = import ../../../pkgs/top-level/aliases.nix;
+  };
+
+  defaultPackages = originalPackages // {
+    all = top: self: pkgs: originalPackages.all top self pkgs // (
+      let callPackage = top.lib.callPackageWith pkgs; in {
+        test-dead = callPackage pkgTestDead { };
+        test-beef = callPackage pkgTestBeef { };
+        test-ate  = callPackage pkgTestAte  { };
+        test-bad  = callPackage pkgTestBad  { };
+        test-food = callPackage pkgTestFood { };
+      });
+  };
+
+  # Only change the foo package which is used explicitly by bar, and
+  # indirectly by baz.
+  quickfixPackages = originalPackages // {
+    all = top: self: pkgs: originalPackages.all top self pkgs // (
+      let callPackage = top.lib.callPackageWith pkgs; in {
+        test-dead = callPackage pkgTestDead { };
+        test-beef = callPackage pkgTestBeef { name = "beef-1.0.2"; };
+        test-ate  = callPackage pkgTestAte  { };
+        test-bad  = callPackage pkgTestBad  { name = "bad-1.0.51"; };
+        test-food = callPackage pkgTestFood { };
+      });
```

Comment:
@danbst, Note that `test-bad` is recompiled with the patched version of `test-ate`, and not the recompiled version of `test-ate`.  This is verified by the line 148, which checks that the dependency names inherited from `test-ate` is `beef-1.0.0` instead of `beef-1.0.2`.

I guess, for testing what you suggest we should copy the `dependency` file from the latest added hash of the `installed` file, and check that the names are the original one, and not the patched one.

With this method, recompiled package depending on `test-food` / `test-ate` would get the renamed version of `test-bad` / `test-beef`, and non recompiled package depending on `test-food` / `test-ate` would get the original version of `test-bad` / `test-beef`.

Note, the way the current patching mechanism is made also guarantee that adding the above test should work, because it only works on one package at a time, thus by recursion inference, all should work out of the box.